server/middleware/users.coffee

```diff
@@ -57,3 +58,54 @@ module.exports =
     yield User.update({ _id: userID }, { $set: { "role": "student" } })
     user = yield User.findById req.user.id
     res.status(200).send(user.toObject({req: req}))
+
+  verifyEmailAddress: wrap (req, res, next) ->
+    user = yield User.findOne({ _id: mongoose.Types.ObjectId(req.params.userID) })
+    [timestamp, hash] = req.params.verificationCode.split(':')
+    if user and req.params.verificationCode is user.verificationCode(timestamp)
+      yield User.update({ _id: user.id }, { emailVerified: true })
+      res.status(200).send({ role: user.toObject().role })
+    else
+      next()
+
+  resetEmailVerifiedFlag: wrap (req, res, next) ->
+    newEmail = req.body.email
+    _id = mongoose.Types.ObjectId(req.body._id)
+    if newEmail
+      user = yield User.findOne({ _id })
+      oldEmail = user.toObject().email
+      if newEmail isnt oldEmail
+        yield User.update({ _id }, { $set: { emailVerified: false } })
+    next()
+
+  sendVerificationEmail: wrap (req, res, next) ->
+    user = yield User.findOne({ _id: mongoose.Types.ObjectId(req.params.userID) })
+    timestamp = (new Date).getTime()
+    if not user
+      res.status(404).send()
+    context =
+      email_id: sendwithus.templates.verify_email
+      recipient:
+        address: user.get('email')
+        name: user.broadName()
+      email_data:
+        name: user.broadName()
+        verify_link: "http://codecombat.com/user/#{user._id}/verify/#{user.verificationCode(timestamp)}"
+    sendwithus.api.send context, (err, result) ->
+    res.status(200).send({})
+
+  teacherPasswordReset: wrap (req, res, next) ->
+    ownedClassrooms = yield Classroom.find({ ownerID: mongoose.Types.ObjectId(req.user.id) })
+    ownedStudentIDs = _.flatten ownedClassrooms.map (c) ->
+      c.toObject().members.map (id) ->
+        id.toString()
+    newPassword = req.body.password
+    studentID = req.params.handle
+    student = yield User.findOne({ _id: mongoose.Types.ObjectId(studentID) })
+    if student.get('emailVerified')
+      return next new errors.Forbidden("Can't reset password for a student that has verified their email address.")
+    if newPassword and studentID in ownedStudentIDs
+      yield User.update({ _id: mongoose.Types.ObjectId(studentID) }, { $set: { passwordHash: User.hashPassword(newPassword) } })
```

Comment:
Can be shortened to `student.update({ $set: { passwordHash: User.hashPassword(newPassword) } })`